File: [1]Programmers/Level1/소수찾기.cpp
// 내가 어렵게 푼 문제!
#include <string>
#include <vector>

using namespace std;

// 소수의 정의 : 1과 자기 자신만으로 나누어 떨어지는 숫자!
// 에라토스테네스 체 문제 !===========================================================
// 2 => 2는 소수이다. 2의 배수는 모두 2가 인수로 들어가기 때문에 소수가 될 수 없다.
// 3 => 3은 소수이다. 3의 배수는 모두 3이 인수로 들어가기 때문에 소수가 될 수 없다.
// 4 => 4는 2의 배수이기 때문에 소수가 될 수 없다.
// 5 => 5는 2와 3의 배수가 아니기 때문에 소수가 될 수 있다. 또한, 5의 배수는 모두 소수가 될 수 없다.
//
// 문제 해결 전략 -> n까지 숫자까지 중 소수를 모두 구하기 위해서 1~n까지 모두 초기화해준다.
//                그 후 소수 판별과 체를 걸러서 검사한다.
// =============================================================================

int solution(int n) {
    int answer = 0;

    vector<bool> primeNum;
    primeNum.resize(n+1);
    fill(primeNum.begin(), primeNum.end(),true);

    for (int i = 2; i < n+1; i++) {
        if (!primeNum.at(i)) {
            continue;
        }else {
            answer++;
            for (int j=1 ; i*j <= n; j++) {
                primeNum.at(i*j) = false;
            }
        }
    }
    return answer;
}

File: Java_Algorithm/src/프로그래머스/Level2/소수찾기.java
package 프로그래머스.Level2;

import java.util.ArrayList;

class Solution_소수찾기 {
    ArrayList<Integer> numberList = new ArrayList<Integer>();

    public int solution(String numbers) {
        int numberMaxLength = numbers.length();
        int depth = 0;
        int answer = 0;
        String[] stringNumberList = numbers.split("");

        for(int r = 1 ; r <= numberMaxLength; r++){
            String[] tempList = new String[r];
            perm(stringNumberList, depth, numberMaxLength, r, tempList);
        }

        for(int i : numberList){
            boolean primeCheck = true;
            for(int j = 2 ; j < i ; j++){
                if( i % j == 0){
                    primeCheck = false;
                    break;
                }
            }
            if(primeCheck){
                answer++;
            }
        }
        return answer;
    }

    public void perm(String[] stringNumberList, int depth, int length,int r, String[] tempList){

        if (depth == r) {
            makeNumber(tempList);
            return;
        }

        for (int i = depth; i < length; i++) {
            swap(stringNumberList, i, depth);
            tempList[depth] = stringNumberList[depth];
            perm(stringNumberList, depth + 1, length, r,tempList);
            swap(stringNumberList, i, depth);
        }

    }

    public void swap(String[] stringNumberList, int i, int depth){
        String temp = stringNumberList[i];
        stringNumberList[i] = stringNumberList[depth];
        stringNumberList[depth] = temp;
    }

    public void makeNumber(String[] tempList){
        String number = "";
        for(int i = 0 ; i < tempList.length ; i++){
            number += tempList[i];
        }
        addNumber(number);
    }

    public void addNumber(String number){
        int num = Integer.parseInt(number);
        if(num == 0 || num == 1){
            return;
        }

        if(!(numberList.contains(num))){
            numberList.add(num);
        }
    }
}


public class 소수찾기 {
    public static void main(String[] args) {
        Solution_소수찾기 s = new Solution_소수찾기();
        String number1 = "17";
        String number2 = "011";

        //System.out.println(number1 + " : " + s.solution(number1));
        System.out.println(number2 + " : " + s.solution(number2));

    }
}
